// oo/model/models/User.js
let Parent = require('./Model');

class User extends Parent {

    constructor() {
        super();
    }

    async init() {
        await super.init();
    }

    getFields(){
        return {
            Username: {type: 'string', required: true},
            Name: {type: 'string', required: true},
            Password: {type: 'string', required: true},
            Admin: {type: 'boolean'},
            Email: {type: 'string', required: true},
            ID: {type: 'string', required: true},
            Address: {type: 'string', required: true, tab: 'Address'},
            ZipCode: {type: 'integer', required: true, tab: 'Address'},
            Locality: {type: 'string', required: true, tab: 'Address'},
            City: {type: 'string', required: true, tab: 'Address'},
            Country: {type: 'string', required: true, tab: 'Address'},
        };
    }

    getUniqueFields(){
        return ['Username', 'Email'];
    }

    static getCheckoutShipFields() {
        return {
            Email: {type: 'string', required: true},
            Name: {type: 'string', required: true},
            ID: {type: 'string', required: true},
            Address: {type: 'string', required: true},
            ZipCode: {type: 'integer', required: true},
            Locality: {type: 'string', required: true},
            City: {type: 'string', required: true},
            Country: {type: 'string', required: true},
            PayMethods: {type: 'set', required: true, linkto: 'paymethods'},
            ShipMethods: {type: 'set', required: true, linkto: 'shipmethods'}
        };
    }

    static getLoginFields() {
        return ['Username', 'Password'];
    }

    static getRegisterFields() {
        let fields = ['Name','Email'];
        return [...this.getLoginFields(), ...fields];
    }

}

module.exports = User;
